### ISSUE TITLE
[C++][Python] pyarrow.dataset.write_dataset does not store fields metadata

### ISSUE THREAD

**Issue by hkad98** [apache/arrow#34067]
### Describe the usage question you have. Please include as many useful details as  possible.

Hello,
I noticed that `pyarrow.dataset.write_dataset` does not store field metadata. Could you please, tell me if it is expected behavior?

I am using pyarrow version 10.0.1. I also tried 11.0.0, and the behavior is the same.

```
import pandas as pd
import pyarrow as pa
import pyarrow.dataset

df = pd.DataFrame({'year': [2020, 2022, 2019, 2021],
                    'n_legs': [2, 4, 5, 100],
                    'animals': ["Flamingo", "Horse", "Brittle stars", "Centipede"]})
schema = pa.schema([
    pa.field('year', pa.int16(), metadata={"123": b"123"}),
    pa.field('n_legs', pa.int64()),
    pa.field('animals', pa.string())],
    metadata={"xyz": b"xyz"})
t = pa.Table.from_pandas(df, schema)
print(t.field(0).metadata)
# prints {b'123': b'123'}

pyarrow.dataset.write_dataset(
        data=t,
        base_dir="xyz",
        schema=t.schema,
        format="arrow",
    )
```
Then load the stored file

```
with pa.memory_map("xyz/part-0.arrow", "r") as source:
        table = pa.ipc.open_file(source).read_all()
print(table.field(0).metadata)
# prints None
print(table.schema.metadata)
# prints {b'xyz': b'xyz', b'pandas': ... }
```

As you can see above, schema metadata are preserved.

### Component(s)

Python

**Comment by westonpace** [comment]
No, there is no good reason for this.  If the input is a table then the dataset writer should be able to retain the field metadata.

The `WriteDataset` API is very generic at the moment.  It takes, as input, a scanner.  That scanner can add new columns or remove columns.  So we can't just grab the input schema.  Today we just grab the schema metadata instead.

We can change the write node to accept a schema, that shouldn't be too difficult. Then perhaps we can create an overload for `FileSystemDataset::Write` that accepts a table as input instead of a scanner.

**Comment by jorisvandenbossche** [comment]
> No, there is no good reason for this. 

Note that this in general depends on the file format (it needs to support a notion of metadata). But indeed, for Arrow (IPC) file format, we should be able to support it.

For Parquet probably as well, but there are some bugs related to this in the Parquet writer itself (not directly related to datasets): https://github.com/apache/arrow/issues/31723

**Comment by jorisvandenbossche** [comment]
And related (possibly duplicate) issue: https://github.com/apache/arrow/issues/33335

**Comment by westonpace** [comment]
Yes, definitely related but I don't think they are duplicates.  #33335 is more about the missing C++ API.  I also agree this will only be possible when writing to formats that support metadata.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 365 days. Please remove the stale label or comment below, or this issue will be closed in 14 days. If this usage question has evolved into a feature request or docs update, please remove the 'Type: usage' label and add the 'Type: enhancement' label instead.
